### PR TITLE
Add Pydantic Data Model to Document Node Attributes Data Structure

### PR DESCRIPTION
**Pull request by ytausch**
Currently, there is no documentation about the data model of the cf-graph that this bot implicitly uses.

This is the first step of improving this:
- add a Pydantic model for the node attributes
- add a CI check that validates the model against the current cf-graph state
- the above step lead to the discovery of invalid feedstocks and node attributes files, which are listed in `test_validate.py`.
- README updates
- add PR template encouraging contributors to update the Pydantic model

Documenting the data model also included discovering some undocumented features of the `meta.yaml` file. I submitted corresponding upstream docs updates (https://github.com/conda/conda-build/pull/5214, https://github.com/conda/conda-build/pull/5215, https://github.com/conda/conda-build/pull/5217, https://github.com/conda/conda-build/pull/5218)

### CODE REVIEW

**Comment by beckermr** [comment]
This PR is great. I do want to point out that I likely won't have the time or expertise in the short-term to maintain this enhancement by myself.

File: .github/workflows/test-model.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
+name: test-model
+
+on:
+  push:
+    branches:
+      - master
+  pull_request: null
```
Comment (0xbe7a) [inline .github/workflows/test-model.yml:7]:
```suggestion
  pull_request:
```

File: .github/workflows/test-model.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  tests:
+    name: tests
+    runs-on: "ubuntu-latest"
+    defaults:
+      run:
+        shell: bash -l {0}
+
+    steps:
+      - name: cancel previous runs
+        uses: styfle/cancel-workflow-action@0.6.0
```
Comment (0xbe7a) [inline .github/workflows/test-model.yml:22]: Why are the native concurrency settings not used here?

File: .github/workflows/test-model.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
+name: test-model
+
+on:
```
Comment (0xbe7a) [inline .github/workflows/test-model.yml:3]: Does it make sense to do run this on push? New errors are probably not due to a code change, but to errors in cf-graphs / the currently running bot code? I think workflow_dispatch / cron triggers are better suited here

Comment (ytausch) [reply .github/workflows/test-model.yml:3]: Agree. I will change it to workflow_dispatch / cron triggers.

File: .github/workflows/test-model.yml (hunk 2 above)
Comment (ytausch) [reply .github/workflows/test-model.yml:22]: I (too blindly) copied this from `tests.yml`. Will change now.

File: .github/workflows/test-model.yml (hunk 1 above)
Comment (ytausch) [reply .github/workflows/test-model.yml:7]: Changing to workflow_dispatch / cron (again, this is a leftover from `tests.yml`, my bad)

File: conda_forge_tick/models/conda_forge_yml.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+from pydantic import BaseModel
+
+
+class BuildPlatform(str, Enum):
+    def __str__(self):
+        return self.value
+
+    """
+    Build platforms prefixed with LEGACY were mentioned in the documentation for the `provider` field in the
+    `conda-forge.yml` file but are no longer supported. A lot of feedstocks (today: ~3000) still use them.
```
[the remark this answers is not in the document]
Comment (ytausch) [reply conda_forge_tick/models/conda_forge_yml.py]: Fixed with [bbb103b](https://github.com/regro/cf-scripts/pull/2239/commits/bbb103b94e6f78ada0299c4f35214912dbc023d9) (also for other instances)

File: .github/workflows/test-model.yml (hunk 2 above)
Comment (beckermr) [reply .github/workflows/test-model.yml]: FWIW, the GHA stuff here was done before github added that feature.

**Comment by ytausch** [comment]
@beckermr
I have noted this here (at the same time as your comment): https://github.com/regro/cf-scripts/blob/3b5d0da379cbdc8a01586ee563ce2cc0ee7aa029/conda_forge_tick/models/conda_forge_yml.py#L18-L22

**Comment by ytausch** [comment]
> The smithy PR will be merged by the end of the week. We should move to using the stuff from smithy now.

I will not extend the current `conda-forge.yml` model definition from now since more fields are not used by the bot. We can either wait for the smithy PR to be merged and then change this here or migrate to using smithy as a dependency in a separate PR. I was thinking about the latter option.

**Comment by beckermr** [comment]
OK sounds good. I'll leave this PR open then and maybe @jaimergp can help me put in the changes needed to use the new smithy items. Thanks a bunch!

**Comment by ytausch** [comment]
Confirmation: `mypy` did not discover any new type issues with the changes of this PR.

**Comment by jaimergp** [comment]
Thanks! This is on my list for the week. I'll get back to you.

**Comment by jaimergp** [comment]
The schema is now in conda-smithy@main if you want to update the PR!

**Comment by ytausch** [comment]
> The schema is now in conda-smithy@main if you want to update the PR!

Yes, I will look into it now :)

**Comment by ytausch** [comment]
> The schema is now in conda-smithy@main if you want to update the PR!

blocked by https://github.com/conda-forge/conda-smithy/issues/1858

**Comment by beckermr** [comment]
The latest smithy is now released. You need v3.32.0

**Comment by ytausch** [comment]
I want to clarify this PR's status. The following tasks are open, waiting to be resolved:

- [ ] https://github.com/conda-forge/conda-smithy/pull/1865 (some node files fail to validate because of this)
- [ ] https://github.com/conda-forge/libitk-feedstock/pull/88 (invalid feedstock with easy fix)
- [ ] https://github.com/conda-forge/conda-smithy/issues/1869 (some node files fail to validate because of this)

Those all deal with special cases of the `conda-forge.yml` validation and only affect the "known bad" feedstock list in the test. Besides that, I consider this PR complete and I will not add any changes related to the issues above. 

/Cc @0xbe7a

File: tests/model/test_validate.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    "sparc-x",  # `conda-forge.yml`.channels is unexpected
+    "jupyter_core",  # `conda-forge.yml`.abi_migration_branches is unexpected, should be moved to `conda-forge.yml`.bot
+    "rerun-sdk",  # unrecognized field `conda-forge.yml`.azure.artifact_retention_days
+    "bamnostic",  # unrecognized field `conda-forge.yml`.build
+    "r-v8",  # unrecognized field `conda-forge.yml`.github.win
+    "python-utils",  # unrecognized field `conda-forge.yml`.dependencies
+    "pyrosm",  # unrecognized option `conda-forge.yml`.build, legacy field `conda-forge.yml`.matrix does not validate
+    "sketchnu",  # `conda-forge.yml`.conda_build.pkg_format may not be None
+    "fd-find",  # invalid `conda-forge.yml`.test_on_native_only value
+    "micromamba",  # invalid value for `conda-forge.yml`.channel_priority
```
Comment (beckermr) [inline tests/model/test_validate.py:90]: This field appears allowed: https://github.com/conda-forge/conda-smithy/blob/main/conda_smithy/schema.py#L673

Is this list of known failures out of date?

Comment (ytausch) [reply tests/model/test_validate.py:90]: I'm currently updating it, one sec :)

Comment (ytausch) [reply tests/model/test_validate.py]: should be up to date now @beckermr

**Pull request merged**